torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -756,6 +769,38 @@ void Reducer::finalize_backward() {
     local_used_work_->wait();
   }
   local_used_maps_reduced_ = false;
+  rpc_context_.clear();
+}
+
+void Reducer::runGradCallbackForVariable(
+    torch::autograd::Variable& variable,
+    GradCallback&& cb) {
+  auto context_ptr = rpc_context_.context_ptr.load();
+  if (context_ptr == nullptr) {
+    cb(variable.grad());
+  } else {
+    // Under distributed autograd
+    context_ptr->runGradCallbackForVariable(variable, std::move(cb));
+  }
+}
+
+void Reducer::RpcContext::set(ContextPtr&& new_context_ptr) {
+  if (!new_context_ptr) {
+    // Not under distributed autograd
+    return;
+  }
+  if (context_ptr.exchange(new_context_ptr.get()) == nullptr) {
+    // Set the shared ptr to the context only if it's set first time.
+    // All call sites should use the same context ptr.
+    // Use an atomic to avoid data race from multiple threads.
+    context_ptr_holder = std::move(new_context_ptr);
```

Comment:
Why do we need an additional atomic here? `context_ptr = std::move(new_context_ptr)` should be thread safe.

Reply:
Maybe I'm misunderstanding your point. `std::share_ptr::operator=()` is not thread-safe. Furthermore, `std::atomic<std::shared_ptr<int>>` doesn't work, because `std::atomic` requires the type be trivially copyable.

Reply:
Nvm my bad, its not thread safe and needs synchronization. Although, it seems like we can use https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic instead? Would this avoid the extra atomic variable?

Reply:
That would work, but I'd rather avoid them since all functions are marked as `deprecated in C++20`. (I don't know why)